#!/usr/bin/python3
# -*- coding : utf-8 -*-
##########################################################
#
#       Filename:   cp_from_gbk_get_cds.py
#         Author:   yujie
#    Description:   cp_from_gbk_get_cds.py
#        Version:   3.0
#           Time:   2022/03/09 15:21:51
#  Last Modified:   2022/12/17 11:20:02
#        Contact:   hi@arcsona.cn
#        License:   GNU General Public License v3.0
#
##########################################################
from Bio import SeqIO
from Bio.Seq import Seq
# from icecream import ic
import argparse
# import linecache
import os
# import pretty_errors
import re
import sys
import time
# import copy
parser = argparse.ArgumentParser(
    add_help=False, usage='\n\
python3 cp_from_gbk_get_cds_V3.0.py -i [gbk dir] -o [out dir]\n\
Each species generates three files of cds,trna and complete sequence.')
optional = parser.add_argument_group('optional')
required = parser.add_argument_group('required')
required.add_argument(
    '-i', '--input',  help='gbk dir path', type=str)
required.add_argument(
    '-o', '--output',  help='output dir path', type=str)
optional.add_argument('-c', '--check', help='create genes via gap,not running by default,input "-c" when using',
                      action='store_true', required=False)
optional.add_argument('-info', '--info', help='show update log and exit',
                      action='store_true', required=False)
optional.add_argument('-h', '--help', action='help',
                      help='show this help message and exit')
args = parser.parse_args()

if args.info:
    print('\næ›´æ–°æ—¥å¿—:')
    print('\t20221217 feat: âœ¨ å¯¹gbkæ–‡ä»¶è¿›è¡Œå»é‡')
    print('\t20221219 ğŸfix(get_gene_note): ä¿®æ”¹æ²¡æœ‰geneæ ‡ç­¾çš„cdsçš„é»˜è®¤ID')
    print('\t20221220 âœ¨feat(main): å†™å…¥å»é‡åçš„ç™»å½•å·')
    sys.exit(0)

# ############################################################################################


def format_fasta(note, seq, num):
    format_seq = ""
    for index, char in enumerate(seq):
        format_seq += char
        # if (index + 1) % num == 0:#å¯ä»¥ç”¨æ¥æ¢è¡Œ
        # format_seq += "\n"
    return note.strip() + "\n" + format_seq + "\n"


def ir(s):  # åå‘äº’è¡¥
    re = s[::-1]  # å­—ç¬¦ä¸²åå‘
    c = ""  # å®šä¹‰å­—ç¬¦ä¸²cæ¥æ”¶äº’è¡¥åºåˆ—
    for i in re:
        if i == 'A':
            c = c + 'T'
        elif i == 'G':
            c = c + 'C'
        elif i == 'T':
            c = c + 'A'
        elif i == 'C':
            c = c + 'G'
    return c


def merge_sequence(ele, complete_seq):  # åˆå¹¶è·å–åˆ°çš„åºåˆ—
    gene_seq = ""
    tmp_list = []  # ä½ç½®åˆ—è¡¨
    for ele1 in ele.location.parts:
        if ele1.strand == (-1):
            # print('minus')
            tmp_list.append(re.findall(
                r'\d+', str(ele1.end))[0])  # å®é™…èµ·ç‚¹,ä»endä¸­å–ä¸ç”¨+1
            tmp_list.append(str(int(re.findall(
                r'\d+', str(ele1.start))[0])+1))  # å®é™…ç»ˆç‚¹,ä»startå–+1
            gene_seq += ir(complete_seq[ele1.start:ele1.end])
        elif ele1.strand == (1):
            # print('plus')
            tmp_list.append(str(int(re.findall(
                r'\d+', str(ele1.start))[0])+1))  # å®é™…èµ·ç‚¹,è¦+1
            tmp_list.append(re.findall(
                r'\d+', str(ele1.end))[0])  # å®é™…ç»ˆç‚¹,ä¸ç”¨+1
            # åˆ‡ç‰‡æ²¡é—®é¢˜,ç´¢å¼•ä»startåˆ°end-1,ä¹Ÿå°±æ˜¯å¯¹åº”start+1åˆ°endçš„åºåˆ—
            gene_seq += complete_seq[ele1.start:ele1.end]
    return tmp_list, gene_seq


def get_complete_note(seq_record):  # è·å–æ•´ä¸ªå®Œæ•´åŸºå› ç»„ID
    try:
        seq_id = ''
        # 20220819 NC_044756.1.gbk voucher Liu HM/CP02 chloroplast  æœ‰ç‰¹æ®Šç¬¦å·ï¼Œéœ€è¦å¤„ç†
        seq_record.description = seq_record.description.replace('/', '_')
        # if seq_record.description.find('chloroplast'):#æœ‰bug,ç”¨stræ ¼å¼åŒ–åå°±æ²¡é—®é¢˜äº†
        # 20220627 if str(seq_record.description).find('chloroplast') -1ä¹Ÿæˆç«‹,åˆ¤æ–­æ—¶ä¸€å®šè¦ä»¥False Trueä¸ºå‡†
        # or seq_record.description.split(',')[-2].split()[-1] == 'chloroplast' or seq_record.description.split(',')[-2].split()[-1] == 'plastid':
        if str(seq_record.description).find('chloroplast') > 0 \
            or str(seq_record.description).find('plastid') > 0 \
                or seq_record.description.split(',')[-2].split()[-1] == 'chloroplast' \
                or seq_record.description.split(',')[-2].split()[-1] == 'plastid':  # gbk å±äº å¶ç»¿ä½“
            if str(seq_record.description).find('chloroplast') > 0:  # 20220819 æ ¹æ®ä¸åŒå…³é”®è¯åˆ†å‰²
                seq_id = seq_record.description.split(
                    'chloroplast')[0].replace(' ', '_').rstrip('_')
            elif str(seq_record.description).find('plastid') > 0:
                seq_id = seq_record.description.split(
                    'plastid')[0].replace(' ', '_').rstrip('_')
            name = seq_record.name
            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"  # chloroplast--å¶ç»¿ä½“
        elif seq_record.description.split(',')[-2].split()[-1] == 'mitochondrion':
            seq_id = seq_record.description.split(
                'mitochondrion')[0].replace(' ', '_').rstrip('_')
            name = seq_record.name
            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"  # mitochondrion--çº¿ç²’ä½“
        else:
            print('Genome Type WARNING! {}!'.format(
                seq_record.description.split(', ')[-2].split()[-1]))
            complete_note = ">" + (seq_record.description.split('chloroplast')
                                   [0]).replace(' ', '_').rstrip('_') + "\n"
    except:  # å¦‚æœé‡åˆ°ä»»ä½•å‡ºé”™
        print('try/except')
        complete_note = ''
        # gbk_type = input('genome type(1:chloroplast;2:mitochondrion): ')
        gbk_type = 1
        if gbk_type == 1:
            seq_id = seq_record.description.split(
                'chloroplast')[0].replace(' ', '_').rstrip('_')  # ç‰©ç§æˆ–æ ·å“å

            if seq_id.startswith('UNVERIFIED:_'):  # å»æ‰ UNVERIFIED:_
                seq_id = seq_id.lstrip('UNVERIFIED:_')
            # å»æ‰Cerion_watlingense_voucher_USNM:1514170_MN904501 ä¸­ å†’å·åçš„å†…å®¹
            if len(seq_id.split(':')) > 1:
                seq_id = seq_id.split(':')[0]

            name = seq_record.name  # è¦ä¹ˆæ˜¯ç™»å½•å·  è¦ä¹ˆæ˜¯æ ·æœ¬

            if seq_id == name:
                seq_id = seq_id
            elif seq_id != name:
                seq_id = seq_id+'_'+name
            complete_note = ">" + seq_id + "\n"

    return complete_note, seq_id


def get_gene_note(file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name):  # è·å–geneçš„idåŠåºåˆ—
    '''
    ä¼ å…¥ä¸Šä¸€ä¸ªåŸºå› ä¿¡æ¯tmp_gene_name,è¿”å›å½“å‰åŸºå› ä¿¡æ¯tmp_gene_name
    OrderedDict(
        [
        ('codon_start', ['1']),
        ('gene', ['rps19']),
        ('product', ['ribosomal protein S19']),
        ('protein_id', ['UKP82335.1']),
        ('transl_table', ['11']),
        ('translation', [
         'MTRSLKKNPFVANHLLRKINKLNTKAEKEIIITWSRASTIIPTMIGHTIAIHNGKEHLPIYITDRMVGHKLGEFSPTLNFRGHAKNDNRSRR'])
        ]
        )
    '''
    if 'gene' not in ele.qualifiers.keys():
        # è¿”å›ä¸Šä¸€ä¸ªåŸºå› ,å¥½ä»å…¶ä»–å‚è€ƒæ‰¾è¿™ä¸ªæ²¡åå­—çš„
        try:
            tmp_gene_name = tmp_gene_name+'_next'
        except:
            tmp_gene_name = input(
                "\n{0}: {1} Previous: {2}. Current: {3}.\nPlease input current gene name:".format(file_no, file_name, tmp_gene_name, ele.location.parts))
    else:
        tmp_gene_name = ele.qualifiers['gene'][0]

    tmp_list, gene_seq = merge_sequence(ele, complete_seq)
    pos_info_in_gene_note = " ["
    for i in range(2*len(ele.location.parts)):
        if i % 2 == 0:
            pos_info_in_gene_note += tmp_list[i]+'..'
        elif i % 2 == 1:
            pos_info_in_gene_note += tmp_list[i]+';'
    pos_info_in_gene_note = pos_info_in_gene_note.rstrip(';')+"]"
    gene_note = ">" + seq_id + pos_info_in_gene_note + " [gene=" + \
        tmp_gene_name + "]" + "\n"  # '>'åçš„æ ¼å¼å’Œå·²æœ‰è„šæœ¬å…¼å®¹
    return gene_note, gene_seq, tmp_gene_name


def gene_name_standardization(gene_name):  # æ ¼å¼åŒ–åŸºå› åå­—,å¯é‡å¤ä½¿ç”¨
    all_gene_list_upper = ['ATP6', 'ATP8', 'CYTB', 'COX1', 'COX2',
                           'COX3', 'ND1', 'ND2', 'ND3', 'ND4', 'ND4L', 'ND5', 'ND6']
    all_gene_list_lower = ['atp6', 'atp8', 'cob', 'cox1', 'cox2',
                           'cox3', 'nad1', 'nad2', 'nad3', 'nad4', 'nad4l', 'nad5', 'nad6']
    if gene_name.upper() in all_gene_list_upper:
        gene_name = gene_name.upper()
    else:
        i = 0
        while i < 13:
            if all_gene_list_lower[i] == gene_name:
                gene_name = all_gene_list_upper[i]
                break
            else:
                i += 1
        if i >= 13:
            print(gene_name)
            print('WARNING!Please check!')
    return gene_name


def get_gene(gbk_file_path, flag, dict_gene_len, file_no):  # è§£ægbkæ–‡ä»¶è·å–cds
    """å®Œæ•´åŸºå› ç»„"""
    file_name = os.path.basename(gbk_file_path)
    seq_record = SeqIO.read(gbk_file_path, "genbank")
    complete_seq = str(seq_record.seq)
    complete_note, seq_id = get_complete_note(seq_record)
    complete_fasta = format_fasta(complete_note, complete_seq, 80)  # 80æ¢è¡Œæœ¬ä¾‹ä¸é‡‡ç”¨
    """geneåºåˆ—"""
    cds_count = 0  # å¯¹cdsæ•°é‡è®¡æ•°
    trna_count = 0
    cds_fasta = ""
    trna_fasta = ""
    list_cds_name = []  # ç»Ÿè®¡cdsç§ç±»ï¼Œè®°å½•é‡å¤
    list_trna_name = []  # ç»Ÿè®¡trnaç§ç±»,è®°å½•é‡å¤
    tmp_gene_name = ''  # ä¸Šä¸€ä¸ªåŸºå› åå­—,ä¸ºå­å‡½æ•°get_gene_note()å‡†å¤‡çš„
    for ele in seq_record.features:
        '''
        CDS
        tRNA
        rRNA
        repeat_region
        '''
        if ele.type == "CDS":
            cds_count += 1
            cds_note, cds_seq, tmp_gene_name = get_gene_note(
                file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name)
            # list_cds_name.append(tmp_gene_name)  # æœ¬æ¬¡çš„åŸºå› åå­— å¤ç”¨
            cds_fasta += format_fasta(cds_note, cds_seq, 70)
            gene_name = tmp_gene_name
            gene_name = gene_name  # gene_name_standardization(gene_name)
            list_cds_name.append(gene_name)  # å­˜å…¥åˆ—è¡¨
            # dict_gene_len[gene_name].append(3*(len(ele.qualifiers['translation'][0])+1))  # cdsåºåˆ—é•¿åº¦
            if (flag):  # eleæœ‰å¯èƒ½æ˜¯trna,è¦ç¡®ä¿å…ˆæ‰¾åˆ°ä¸€ä¸ªcdsåæ‰èƒ½é€€å‡º,æ‰€ä»¥æ”¾ä¸Šé¢ifçš„ä¸‹ä¸€çº§
                break
        elif ele.type == 'tRNA':
            trna_count += 1
            trna_note, trna_seq, tmp_gene_name = get_gene_note(
                file_no, file_name, ele, complete_seq, seq_id, tmp_gene_name)
            trna_fasta += format_fasta(trna_note, trna_seq, 70)
            gene_name = tmp_gene_name
            gene_name = gene_name  # gene_name_standardization(gene_name)
            list_trna_name.append(gene_name)  # å­˜å…¥åˆ—è¡¨

    s = '{2}: {0} has {1} CDS'.format(file_name, cds_count, file_no)
    if cds_count == 0:
        # --------There may be no comments--------'.format(
        s = '{2}: {0} has {1} CDS'.format(file_name, cds_count, file_no)
        print(s.ljust(50), '----------There may be no comments----------')
    elif cds_count != 0:
        # 20220811 è¾“å‡ºå·¦å¯¹é½  str.ljust(50)  è¾¾åˆ°50ä¸ªå­—ç¬¦å®½åº¦
        print(s.ljust(50), '+', 'list_cds_name')

    return file_name, seq_id, complete_fasta, cds_fasta, cds_count, list_cds_name,  trna_fasta, trna_count, list_trna_name, dict_gene_len, s


def create_gene_by_gap(dict_missing_gene, dict_gene_len, cds_file_path):  # ç”¨gapæ„é€ æ²¡æœ‰çš„åŸºå› 
    for i in dict_missing_gene.keys():
        cds_fasta = ''
        for j in dict_missing_gene[i]:
            ave = round(sum(dict_gene_len[j]) /
                        len(dict_gene_len[j]))  # è¯¥åŸºå› å¹³å‡é•¿åº¦
            gene_note = (i+' [0..0]'+' [gene={}]').format(j)
            gene_seq = ave*'-'
            cds_fasta += format_fasta(gene_note, gene_seq, 70)
        print(cds_fasta)
        file_name = (i.split('_')[-2]+'_' +
                     i.split('_')[-1]+'.1').lstrip('>')
        with open(cds_file_path, 'ab+') as f_cds:
            f_cds.write(cds_fasta.encode())


if __name__ == '__main__':
    #################################################################
    # æ ¼å¼åŒ–æˆ2016-03-20 11: 45: 39å½¢å¼
    begin_time = time.time()
    start_time = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
    print('Start Time : {}'.format(start_time))
    #################################################################
    print('\n')
    if not os.path.exists(args.output):
        os.makedirs(args.output)
    if (not os.path.exists(os.path.join(args.output, 'trna'))):
        os.makedirs(os.path.join(args.output, 'trna'))
    if (not os.path.exists(os.path.join(args.output, 'cds'))):
        os.makedirs(os.path.join(args.output, 'cds'))
    if (not os.path.exists(os.path.join(args.output, 'complete'))):
        os.makedirs(os.path.join(args.output, 'complete'))

    """ç»Ÿè®¡åˆå§‹åŒ–"""
    dict_missing_gene = {}  # æ¯ä¸ªæ–‡ä»¶ä¸­ç¼ºå¤±çš„åŸºå› ç»Ÿè®¡,æ€» å­—å…¸
    dict_gene_len = {}  # ç»Ÿè®¡æ¯ä¸ªåŸºå› åœ¨ä¸åŒç‰©ç§ä¸­çš„é•¿åº¦,å–å¹³å‡
    list_seq_id_de_duplication = []  # å»é‡çš„ç‰©ç§
    list_unique_accession = []  # å»é‡çš„ç™»å½•å·
    """åˆå§‹åŒ–"""
    dict_file_cds_count = {}  # æ¯ä¸ªæ–‡ä»¶ä¸­cdsè®¡æ•°
    file_list = [x for x in os.listdir(
        args.input) if os.path.isfile(os.path.join(args.input, x))]
    file_list.sort()  # key=lambda x: int(x.split('.')[0])) #æ ¹æ®æ–‡ä»¶åä¸­çš„æ•°å­—
    """ä¸»ç¨‹åº"""
    file_no = 0
    for file in file_list:
        file_no += 1
        gbk_file_path = os.path.join(args.input, file)
        file_name, seq_id, complete_fasta, cds_fasta, cds_count, list_cds_name,  \
            trna_fasta, trna_count, list_trna_name, dict_gene_len, s = get_gene(
                gbk_file_path, False, dict_gene_len, file_no)
        dict_file_cds_count[file_name] = cds_count  # æ¯ä¸ªæ–‡ä»¶ä¸­cdsè®¡æ•°
        '''
        20221217å¯¹gbkæ–‡ä»¶å»é‡
        '''
        seq_id_content = seq_id.split('_')
        if seq_id.find('NC_') > 0:
            species = '_'.join(seq_id_content[:-2])
            accession = '_'.join(seq_id_content[-2:])
        else:
            species = '_'.join(seq_id_content[:-1])
            accession = '_'.join(seq_id_content[-1:])
        if species not in list_seq_id_de_duplication:
            list_seq_id_de_duplication.append(species)
            list_unique_accession.append(accession)
            """å†™å…¥æ–‡ä»¶"""
            with open((os.path.join(args.output, 'complete', seq_id+'.fasta')), 'wb') as f_complete, \
                open((os.path.join(args.output, 'cds', 'cds_'+file_name.rstrip('.gbk')+'.fasta')), 'wb') as f_cds,\
                    open((os.path.join(args.output, 'trna', 'trna_'+file_name.rstrip('.gbk')+'.fasta')), 'wb') as f_trna:
                f_complete.write(complete_fasta.encode())
                f_cds.write(cds_fasta.encode())
                f_trna.write(trna_fasta.encode())
    print('{} left after removing duplicates'.format(len(list_unique_accession)))
    with open((os.path.join(args.output, 'list_tre')), 'w') as list_handle:
        for i in list_unique_accession:
            list_handle.write(i+'\n')
    print('\n')
    ###############################################################
    end_time = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
    print('End Time : {}'.format(end_time))
    print('Already Run {}s'.format(time.time()-begin_time))
    print('Done')
    ###############################################################
